lib/vtls/unitytls.c

```diff
@@ -0,0 +1,617 @@
+      failf(data, "Error reading client cert file %s", ssl_cafile);
+      return CURLE_SSL_CERTPROBLEM;
+    }
+  }
+
+  /* Load the client private key */
+  if(SSL_SET_OPTION(key)) {
+    connssl->pk = unitytls_key_parse_pem_from_file(SSL_SET_OPTION(key), SSL_SET_OPTION(key_passwd), &err);
+    if(!connssl->pk || err.code != UNITYTLS_SUCCESS) {
+      failf(data, "Error reading private key %s", SSL_SET_OPTION(key));
+      return CURLE_SSL_CERTPROBLEM;
+    }
+  }
+  else {
+    connssl->pk = NULL;
+  }
+
+  /* We don't support CRL */
+  if(ssl_crlfile) {
+    failf(data, "UnityTls does not suppport crl");
+  }
+
+  /* Create and configure context */
+  switch(SSL_CONN_CONFIG(version)) {
+    case CURL_SSLVERSION_DEFAULT:
+      protocol_range = unitytls->UNITYTLS_TLSCTX_PROTOCOLRANGE_DEFAULT;
+      break;
+    case CURL_SSLVERSION_TLSv1_0:
+      protocol_range.max = protocol_range.min = UNITYTLS_PROTOCOL_TLS_1_0;
+      break;
+    case CURL_SSLVERSION_TLSv1_1:
+      protocol_range.max = protocol_range.min = UNITYTLS_PROTOCOL_TLS_1_1;
+      break;
+    case CURL_SSLVERSION_TLSv1_2:
+      protocol_range.max = protocol_range.min = UNITYTLS_PROTOCOL_TLS_1_2;
+      break;
+    default:
+      failf(data, "Unrecognized/unsupported parameter passed via CURLOPT_SSLVERSION");
+      return CURLE_SSL_CONNECT_ERROR;
+  }
+
+  connssl->ctx = unitytls->unitytls_tlsctx_create_client(protocol_range, callbacks, hostname, strlen(hostname), &err);
+  unitytls->unitytls_tlsctx_set_certificate_callback(connssl->ctx, on_certificate_request, connssl, &err);
+  unitytls->unitytls_tlsctx_set_x509verify_callback(connssl->ctx, on_verify, connssl, &err);
+  if(err.code != UNITYTLS_SUCCESS) {
+    failf(data, "Error creating and configuring untiytls context: %i", err.code);
+    return CURLE_SSL_CONNECT_ERROR;
+  }
+
+  connssl->conn = conn;
+  connssl->sockfd = conn->sock[sockindex];
+  connssl->connecting_state = ssl_connect_2;
+
+  /* give application a chance to interfere with SSL set up. */
+  if(data->set.ssl.fsslctx) {
+    CURLcode result = (*data->set.ssl.fsslctx)(data, connssl->ctx, data->set.ssl.fsslctxp);
+    if(result != CURLE_OK) {
+      failf(data, "error signaled by ssl ctx callback");
+      return result;
+    }
+  }
+
+  return CURLE_OK;
+}
+
+static CURLcode unitytls_connect_step2(struct Curl_easy* data, struct ssl_connect_data* connssl)
+{
+  unitytls_errorstate err = unitytls->unitytls_errorstate_create();
+  unitytls_x509verify_result verifyresult = unitytls->unitytls_tlsctx_process_handshake(connssl->ctx, &err);
+  CURLcode result = CURLE_OK;
+
+  if (err.code == UNITYTLS_USER_WOULD_BLOCK) {
+    return CURLE_OK;  /* all fine but no state change yet */
+  }
+
+  if(verifyresult != UNITYTLS_X509VERIFY_SUCCESS) {
+    if(verifyresult == UNITYTLS_X509VERIFY_FATAL_ERROR)
+      failf(data, "Cert handshake failed. verify result: UNITYTLS_X509VERIFY_FATAL_ERROR. error state: %i", err.code);
```

Comment:
This is most likely not a verification error. Maybe use a different return code and not CURLE_PEER_FAILED_VERIFICATION;

Reply:
I agree. I had this returning CURLE_SSL_CONNECT_ERROR in an earlier version, but this made the curl tests less happy. Do you think we should stick to CURLE_SSL_CONNECT_ERROR  anyways?

Reply:
not sure, maybe we need to figure out when curl expect a versification error and when it doesn't. it's possible some of the fatal errors we have are actually verification errors.

Reply:
I reconsidered and went with `CURLE_SSL_CONNECT_ERROR` for `UNITYTLS_X509VERIFY_FATAL_ERROR` now again.

This makes test "0311...[HTTPS wrong subjectAltName but right CN]" fail with "curl returned 35, when expecting 51" since unitytls does not recognize this kind of error as a verification failure. Interestingly, there is a lot of special handling code about this particular piece of verification in curl's openssl implementation while there is nothing in the corresponding mbedtls implementation. (although admittedly I haven't jumped through the necessary hoops to run those tests with mbedtls)
The way our verification code translation works for mbedtls, it seems that we cannot fix this problem for unitytls' mbed backend (we would never give out `UNITYTLS_X509VERIFY_FATAL_ERROR` for any kind of verification error).